### PR TITLE
proxy: port ranges

### LINKED ISSUE #4601: Proxy device: port range
# Required information

 * Distribution: ubuntu
 * Distribution version: 18.04
 * The output of "lxc info" or if that fails:
```
config:
  core.https_address: 192.168.88.100:8443
  core.trust_password: true
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_stop_priority
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
- restrict_devlxd
- migration_pre_copy
- infiniband
- maas_network
- devlxd_events
- proxy
- network_dhcp_gateway
- file_get_symlink
- network_leases
- unix_device_hotplug
- storage_api_local_volume_handling
- operation_description
- clustering
- event_lifecycle
- storage_api_remote_volume_handling
- nvidia_runtime
api_status: stable
api_version: "1.0"
auth: trusted
public: false
auth_methods:
- tls
environment:
  addresses:
  - 192.168.88.100:8443
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
...
    -----END CERTIFICATE-----
  certificate_fingerprint: [masked]
  driver: lxc
  driver_version: 3.0.0
  kernel: Linux
  kernel_architecture: x86_64
  kernel_version: 4.15.0-20-generic
  server: lxd
  server_pid: 1785
  server_version: 3.0.0
  storage: btrfs
  storage_version: "4.4"
  server_clustered: true
  server_name: lxdhost
```

# Issue description

Proxy device port ranges possibly not implemented?

# Steps to reproduce

```
lxc launch ubuntu:18.04 c1
lxc exec c1 -- apt install nginx
lxc config device add c1 http proxy listen=tcp:192.168.88.100:80 connect=tcp:127.0.0.1:80 bind=host
```

using the browser to get 192.168.88.100:80 results in the nginx welcome page. continuing with

```
lxc config device remove c1 http 
lxc config device add c1 http proxy listen=tcp:192.168.88.100:70-90 connect=tcp:127.0.0.1:70-90 bind=host
```

returns `Device http added to c1` but the port range doesnt work. So either the syntax is different (and I couldn't find it), or this isn't implemented. Implementing this is IMO quite important for the usability of the proxy device, some services, i.e. ftp, require a dedicated range of ports (in case of ftp for passive transfers). Other uses may be i.e. in p2p networking, etc.

### PR DESCRIPTION
**Pull request by brauner**
Closes #4601.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

File: doc/containers.md [hunk 1]
```diff
@@ -438,7 +438,7 @@ connect     | string    | -                 | yes       | The address and port t
 bind        | string    | host              | no        | Which side to bind on (host/container)
 
 ```
-lxc config device add <container> <device-name> proxy listen=<type>:<addr>:<port> connect=<type>:<addr>:<port> bind=<host/container>
+lxc config device add <container> <device-name> proxy listen=<type>:<addr>:<port>[-<port>][,<port>] connect=<type>:<addr>:<port> bind=<host/container>
```
[the remark this answers is not in the document]
Comment (stgraber) [reply doc/containers.md]: Agreed, the most common case for this feature is to map a range of ports to the same range of ports inside the container, so say:

`lxc config device add <container> <device-name> proxy listen=udp:1.2.3.4:20000-30000 connect=udp:127.0.0.1:20000-30000`

The other use case we could/should support is multiple ports to single port:

`
lxc config device add <container> <device-name> proxy listen=udp:1.2.3.4:0-65534 connect=udp:127.0.0.1:1194
`

The first example would be what I'd expect for a FTP type use case, the latter would be for a VPN server which binds all ports on a given IP and send them all to the VPN port (1194).

So the listener side can have a list of arbitrary ports and ranges, the connect side can either be a single port (gets all traffic) or must be a range of ports of the same size as the listening side.

**Comment by stgraber** [comment]
@brauner so we need to sort out what @killua-eu and I discussed in-line and with that done, add two tests to check that everything's good, I'd suggest:
 - UDP port range to single UDP port
 - TCP set of ports to different set of ports (but same port count)

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
Found one issue with the relay code which I fixed but I'm still hitting much more failures than would be expected... I've started playing with using `socat` rather than `nc` as an alternative that better lets us configure sending/receiving behavior (read from stdin then disconnect) but I couldn't get it to work well and it's getting late now so I'll let you take another look at potential issues :)

**Comment by brauner** [comment]
jenkins: test this please

**Pull request merged**